fastcrypto/src/vdf/wesolowski.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::error::FastCryptoError::{InvalidInput, InvalidProof};
+use crate::error::FastCryptoResult;
+use crate::groups::class_group::{Discriminant, QuadraticForm};
+use crate::groups::{ParameterizedGroupElement, UnknownOrderGroupElement};
+use crate::hash::{HashFunction, Sha256};
+use crate::vdf::VDF;
+use num_bigint::{BigInt, Sign};
+use num_integer::Integer;
+use num_prime::nt_funcs::is_prime;
+use std::cmp::min;
+use std::ops::Neg;
+
+/// Size of the random prime modulus B used in proving and verification.
+const B_BITS: usize = 264;
+
+/// An implementation of the Wesolowski VDF construction (https://eprint.iacr.org/2018/623) over a
+/// group of unknown order.
+pub struct WesolowskiVDF<G: ParameterizedGroupElement + UnknownOrderGroupElement> {
+    group_parameter: G::ParameterType,
+    iterations: u64,
+}
+
+impl<G: ParameterizedGroupElement + UnknownOrderGroupElement> WesolowskiVDF<G> {
+    /// Create a new VDF using the group defined by the given group parameter. Evaluating this VDF
+    /// will require computing `2^iterations * input` which requires `iterations` group operations.
+    fn new(group_parameter: G::ParameterType, iterations: u64) -> Self {
+        Self {
+            group_parameter,
+            iterations,
+        }
+    }
+}
+
+impl<G: ParameterizedGroupElement<ScalarType = BigInt> + UnknownOrderGroupElement> VDF
+    for WesolowskiVDF<G>
+{
+    type InputType = G;
+    type OutputType = G;
+    type ProofType = G;
+
+    fn evaluate(&self, input: &G) -> FastCryptoResult<(G, G)> {
+        if !input.has_group_parameter(&self.group_parameter) {
+            return Err(InvalidInput);
+        }
+
+        if self.iterations == 0 {
+            return Ok((input.clone(), G::zero(&self.group_parameter)));
+        }
+
+        let mut output = input.clone();
+        for _ in 0..self.iterations {
```

Comment:
iterations = T when in the protocol it says 2^T?

Reply:
Yes, exactly. It's documented in the constructor for the Wesolowski construction, but I'll add a comment here also.